### DISCUSSION TITLE
About Time and Space dependent diffusivity and source terms

### DISCUSSION THREAD

**Discussion by elternativeht** [zwicker-group/py-pde#436] [Q&A] [answered]
Dr. Zwicker,

First of all, kudos for your hard work and this nice package. It has been running great from my end and I cannot emphasize the appreciation.

I am recently running into the issue of 1D diffusion problem where we have both time- and space-varying diffusivity and space- and time-dependent source term. I read through the previous Q&A sessions. In #323 you referred #229 and said that using the same method we can implement time-dependent diffusivity method. Also , in #366 you suggested we could do the same thing towards the source term.

My understanding is for solely space-dependent diffusivity/source terms, I would pre-compute the space-dependent diffusivity (either via explicit expressions or direct computation of all the data in each grid point) and include this diffusivity as `consts` in PDE expressions. However, I do not have good clue of implementing time-dependent diffusivity into `PDE` class using this method, as in my case I will not have explicit expressions on this time dependence. 

In my case, the diffusivity will be changing with space and might change with time (depending on the time scale) and the source term $\dot{m}$ changes with $C$ and $\frac{\partial {C}}{\partial{t}}$. Do you think I would have to implement my own PDE class, or it is possible to stick with current PDE class? If the former, could you give out a high-level approach how I should implement it?

Thank you in advance for your precious time and kind help!

Sincerely regards,

**Comment by david-zwicker** [comment]
The `PDE` class supports explicit dependencies on time `t` and spatial coordinates. Moreover, you can use pre-computed fields in the `PDE` class. With this most cases of spatio-temporal dependence should be possible to implement. I don't see any reason why you would need a custom class at this point, but this might depend on the exact expressions in you PDE.

**Reply by elternativeht** [reply]
Thank you for your prompt reply. I managed to set up the simplified equation. Do you think current package support a field that is a function of potentially the state partial derivative of time? My understanding is no so I wonder if there is any possibility to implement that in the future as well? Thanks again!

**Reply by david-zwicker** [reply, marked as answer]
I'm not sure I understand your problem. Could you give an example for an equation where this would be necessary? It sounds a bit as if you want an implicit definition of the evolution rate, which might not have a unique solution and will be almost impossible to implement generically.

**Reply by elternativeht** [reply]
Thank you Dr. Zwicker! Sorry I was out of internet for a few weeks until now. Yes I do have a more complicated version of evolution rate, but we decided to go for the simpler version of the equation for now and try to get some deliverables first. Thank you again and I appreciate your devotion to the project!
